### PR TITLE
support search issues pagination

### PR DESCRIPTION
**Pull request by lunny**
This adds pagination support to search results, next PR will include template changes to allow for UI to utilize this pagination.

### CODE REVIEW

**Review by techknowlogick — APPROVED** [review]

File: routers/api/v1/repo/issue.go [hunk 1]
```diff
@@ -189,9 +189,18 @@ func SearchIssues(ctx *context.APIContext) {
...
 		keyword = ""
 	}
+	// this api is also used in UI,
+	// so the default limit is set to fit UI needs
+	limit := ctx.FormInt("limit")
+	if limit == 0 {
+		limit = setting.UI.IssuePagingNum
+	} else if limit > setting.API.MaxResponseItems {
+		limit = setting.API.MaxResponseItems
+	}
```
Comment (delvh) [inline routers/api/v1/repo/issue.go:199]: :raised_eyebrow:
Sounds like a bug report waiting to happen.

File: routers/api/v1/repo/issue.go [hunk 2]
```diff
@@ -272,12 +272,14 @@ func SearchIssues(ctx *context.APIContext) {
 			return
 		}
 
-		issuesOpt.ListOptions = db.ListOptions{
-			Page: -1,
-		}
-		if filteredCount, err = issues_model.CountIssues(ctx, issuesOpt); err != nil {
-			ctx.Error(http.StatusInternalServerError, "CountIssues", err)
-			return
+		if filteredCount == 0 {
```
Comment (delvh) [inline routers/api/v1/repo/issue.go:275]: Under what circumstances is `filteredCount` `0` and why should we return the total number of issues then?
Can't we just delete this whole branch now?

File: routers/api/v1/repo/issue.go [hunk 3]
```diff
@@ -481,12 +482,14 @@ func ListIssues(ctx *context.APIContext) {
...
+			issuesOpt.ListOptions = db.ListOptions{
+				Page: -1,
+			}
+			if filteredCount, err = issues_model.CountIssues(ctx, issuesOpt); err != nil {
+				ctx.Error(http.StatusInternalServerError, "CountIssues", err)
+				return
+			}
```
Comment (delvh) [inline routers/api/v1/repo/issue.go:492]: See above

File: routers/web/repo/issue.go [hunk 4]
```diff
@@ -2341,13 +2342,22 @@ func SearchIssues(ctx *context.Context) {
 	var issues []*issues_model.Issue
 	var filteredCount int64
 
+	// this api is also used in UI,
```
Comment (delvh) [inline routers/web/repo/issue.go:2345]: this is no longer an API

File: routers/web/repo/issue.go [hunk 5]
```diff
@@ -2431,12 +2432,14 @@ func SearchIssues(ctx *context.Context) {
 			return
 		}
 
-		issuesOpt.ListOptions = db.ListOptions{
-			Page: -1,
-		}
-		if filteredCount, err = issues_model.CountIssues(ctx, issuesOpt); err != nil {
-			ctx.Error(http.StatusInternalServerError, "CountIssues", err.Error())
-			return
+		if filteredCount == 0 {
```
Comment (delvh) [inline routers/web/repo/issue.go:2435]: See above

File: routers/web/repo/issue.go [hunk 6]
```diff
@@ -2592,12 +2595,14 @@ func ListIssues(ctx *context.Context) {
 			return
 		}
 
-		issuesOpt.ListOptions = db.ListOptions{
-			Page: -1,
-		}
-		if filteredCount, err = issues_model.CountIssues(ctx, issuesOpt); err != nil {
-			ctx.Error(http.StatusInternalServerError, err.Error())
-			return
+		if filteredCount == 0 {
```
Comment (delvh) [inline routers/web/repo/issue.go:2598]: See above

**Comment by wxiaoguang** [comment]
Stale?

File: routers/api/v1/repo/issue.go (hunk 1 above)
Comment (lunny) [reply routers/api/v1/repo/issue.go:199]: These lines just moved from line 222 - 230

**Comment by lunny** [comment]
After finished this PR, I realize we should not support pagination for a search engine. The next page maybe better.

**Comment by techknowlogick** [comment]
For anyone coming to this PR in the future as in the linked issue, this PR is not suitable for pagination due to other issues.

**Comment by lunny** [comment]
I'm writing a new PR which will replace this one and try to fix #24662

**Pull request closed without merging**